### PR TITLE
rustc: Fix joint-ness of stringified token-streams

### LINKED ISSUE #50700: Tokens in an attribute macro receive wrong spacing
Originally reported by @CodeSandwich as https://github.com/dtolnay/syn/issues/425.

We are seeing a functionlike procedural macro invocation `m![< -]` receive the expected spacing (Alone for both tokens) but preprocessing the same input with an attribute macro `#[attribute] m![< -]` receives incorrect spacing -- Joint for the `<` token even though it is followed by whitespace. Strangely using parentheses appears to fix the spacing so that `#[attribute] m!(< -)` correctly has Alone for both tokens.

### Repro script

```sh
#!/bin/sh

cargo new --lib repro
cargo new --lib repro_macros

echo >repro/src/main.rs '
#![feature(proc_macro)]

extern crate repro_macros;
use repro_macros::{tokens, nothing};

// alone + alone
tokens![< -];

// joint + alone
#[nothing]
asdf![< -];

// alone + alone
#[nothing]
asdf!(< -);

fn main() {}
'

echo >>repro/Cargo.toml '
repro_macros = { path = "../repro_macros" }
'

echo >repro_macros/src/lib.rs '
#![feature(proc_macro)]

extern crate proc_macro;
use proc_macro::TokenStream;

#[proc_macro]
pub fn tokens(input: TokenStream) -> TokenStream {
    println!("{:#?}", input);
    TokenStream::empty()
}

#[proc_macro_attribute]
pub fn nothing(_: TokenStream, input: TokenStream) -> TokenStream {
    println!("{:#?}", input);
    TokenStream::empty()
}
'

echo >>repro_macros/Cargo.toml '
[lib]
proc-macro = true
'

cargo build --manifest-path repro/Cargo.toml
```

<details><summary>Output</summary>

```
TokenStream [
    Op {
        op: '<',
        spacing: Alone,
        span: #0 bytes(116..117)
    },
    Op {
        op: '-',
        spacing: Alone,
        span: #0 bytes(118..119)
    }
]
TokenStream [
    Term {
        sym: asdf,
        span: #0 bytes(7615976..7615980)
    },
    Op {
        op: '!',
        spacing: Alone,
        span: #0 bytes(0..0)
    },
    Group {
        delimiter: Parenthesis,
        stream: TokenStream [
            Op {
                op: '<',
                spacing: Joint,
                span: #0 bytes(0..0)
            },
            Op {
                op: '-',
                spacing: Alone,
                span: #0 bytes(0..0)
            }
        ],
        span: #0 bytes(0..0)
    },
    Op {
        op: ';',
        spacing: Alone,
        span: #0 bytes(0..0)
    }
]
TokenStream [
    Term {
        sym: asdf,
        span: #0 bytes(192..196)
    },
    Op {
        op: '!',
        spacing: Alone,
        span: #0 bytes(196..197)
    },
    Group {
        delimiter: Parenthesis,
        stream: TokenStream [
            Op {
                op: '<',
                spacing: Alone,
                span: #0 bytes(198..199)
            },
            Op {
                op: '-',
                spacing: Alone,
                span: #0 bytes(200..201)
            }
        ],
        span: #0 bytes(197..202)
    },
    Op {
        op: ';',
        spacing: Alone,
        span: #0 bytes(202..203)
    }
]
```
</details>

<br>

@alexcrichton

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit fixes `StringReader`'s parsing of tokens which have been stringified
through procedural macros. Whether or not a token tree is joint is defined by
span information, but when working with procedural macros these spans are often
dummy and/or overridden which means that they end up considering all operators
joint if they can!

The fix here is to track the raw source span as opposed to the overridden span.
With this information we can more accurately classify `Punct` structs as either
joint or not.

Closes #50700

### CODE REVIEW

File: src/libsyntax/parse/lexer/mod.rs [hunk 1]
```diff
@@ -121,6 +132,7 @@ impl<'a> StringReader<'a> {
             sp: self.peek_span,
         };
         self.advance_token()?;
+        self.span_src_raw = self.peek_span_src_raw;
```
Comment (matklad) [inline src/libsyntax/parse/lexer/mod.rs]: Heh, I've just spend like an hour perusing this line, and I came to the conclusion that:

* this code is buggy
* the code is observationaly correct

The code is buggy because, while `span` and `token` point to *this* token, `span_src` points to the *next* token. In other words, `span_src_raw` and `peek_span_src_raw` are just equal all the time, because, first `advance_token` sets `peek_span_src_raw`, and then we assign `peek_span_src_raw` to `span_src_raw`.

However, this bug does not show up, because we only use peek span to check for jointness. And if spans A and B are adjacent, than spans *following* A and B are adjacent as well.

In other words, for `++ ` we don't check that first `+` is adjacent to the next, we check that the token following the first `+` (the second `+`) is adjacent to the token, following the second `+` (whitespace).

Similarly, for `+ + `  we actually check that *whitespace* following `+` is not adjacent.
